haystack/schema.py

```diff
@@ -549,23 +566,24 @@ def no_answer(self) -> Optional[bool]:
         return no_answer
 
     def to_dict(self):
-        return asdict(self)
+        return asdict(self, dict_factory=_dict_factory)
 
     @classmethod
-    def from_dict(cls, dict: dict):
-        # backward compatibility for old labels using answers with document_id instead of document_ids
+    def from_dict(cls, dict: Dict):
         answer = dict.get("answer")
-        if answer and "document_id" in answer:
-            dict = dict.copy()
+        if answer and isinstance(answer, builtins.dict):
             dict["answer"] = Answer.from_dict(dict["answer"])
-        return _pydantic_dataclass_from_dict(dict=dict, pydantic_dataclass_type=cls)
+        doc = dict.get("document")
+        if isinstance(doc, builtins.dict):
```

Comment:
```suggestion
        if isinstance(doc, Dict):
```

Reply:
good catch! I wasn't aware before that `Dict` as works with `isinstance`. For some reason I thought it only worked with builtins.